File: ios/Utils/VisionAvailability.swift
import Foundation
import NitroModules

enum VisionAvailability {
  static var supportsBackgroundRemoval: Bool {
    if #available(iOS 17.0, *) {
#if targetEnvironment(simulator)
      return false
#else
      return true
#endif
    }
    return false
  }

  static let supportsImageClassification = true

  /// Modern `RecognizeTextRequest` (Swift concurrency) requires iOS 18+.
  static var supportsTextRecognition: Bool {
    if #available(iOS 18.0, *) {
      return true
    }
    return false
  }

  static var backgroundRemovalUnavailableReason: String {
    if #available(iOS 17.0, *) {
#if targetEnvironment(simulator)
      return "Background removal requires a physical iOS device; the iOS Simulator cannot run Vision foreground masking."
#endif
      return "Background removal is unavailable on this device."
    }
    return "Background removal requires iOS 17 or newer."
  }

  static var textRecognitionUnavailableReason: String {
    "Text recognition requires iOS 18 or newer (Vision RecognizeTextRequest)."
  }

  static func requireBackgroundRemoval() throws {
    guard supportsBackgroundRemoval else {
      throw RuntimeError(backgroundRemovalUnavailableReason)
    }
  }

  static func requireTextRecognition() throws {
    guard supportsTextRecognition else {
      throw RuntimeError(textRecognitionUnavailableReason)
    }
  }
}

File: android/src/main/java/com/margelo/nitro/nitrovisionkit/utils/VisionAvailability.kt
package com.margelo.nitro.nitrovisionkit

import android.content.Context
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability

internal object VisionAvailability {
  const val supportsImageClassification: Boolean = true

  fun supportsBackgroundRemoval(context: Context): Boolean =
    playServicesAvailable(context)

  /** Unbundled ML Kit OCR needs Play Services (same as subject segmentation). */
  fun supportsTextRecognition(context: Context): Boolean =
    playServicesAvailable(context)

  fun backgroundRemovalUnavailableReason(context: Context): String {
    val result = playServicesResult(context)
    if (result == ConnectionResult.SUCCESS) {
      return "Background removal is unavailable on this device."
    }
    return GoogleApiAvailability.getInstance().getErrorString(result)
      ?: "Background removal requires Google Play Services."
  }

  fun textRecognitionUnavailableReason(context: Context): String {
    val result = playServicesResult(context)
    if (result == ConnectionResult.SUCCESS) {
      return "Text recognition is unavailable on this device."
    }
    return GoogleApiAvailability.getInstance().getErrorString(result)
      ?: "Text recognition requires Google Play Services."
  }

  fun requireBackgroundRemoval(context: Context) {
    if (!supportsBackgroundRemoval(context)) {
      throw RuntimeException(backgroundRemovalUnavailableReason(context))
    }
  }

  fun requireTextRecognition(context: Context) {
    if (!supportsTextRecognition(context)) {
      throw RuntimeException(textRecognitionUnavailableReason(context))
    }
  }

  private fun playServicesAvailable(context: Context): Boolean =
    playServicesResult(context) == ConnectionResult.SUCCESS

  private fun playServicesResult(context: Context): Int =
    GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context)
}
